third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs

```diff
@@ -0,0 +1,1453 @@
+                    if src_ty.is_reference() {
+                        let label =
+                            state.make_local(*src, code_offset, (src_qualifier_offset + i) as u8);
+                        let descendant = dest_labels[dest];
+                        self.check_and_add_edge(
+                            state,
+                            label,
+                            BorrowEdge::new(
+                                BorrowEdgeKind::Call(oper.clone()),
+                                dest_ty.is_mutable_reference(),
+                                Some(loc.clone()),
+                                descendant,
+                            ),
+                            alive,
+                        )
+                    }
+                }
+            }
+        }
+    }
+
+    /// Process a MoveFrom instruction.
+    fn move_from(
+        &self,
+        state: &LifetimeState,
+        id: AttrId,
+        dest: TempIndex,
+        resource: &QualifiedInstId<StructId>,
+        src: TempIndex,
+        alive: &LiveVarInfoAtCodeOffset,
+    ) {
+        self.check_read_local(state, id, src, ReadMode::Argument, alive);
+        self.check_write_local(state, id, dest, alive);
+        if let Some(label) = state.label_for_global_with_descendants(resource) {
+            self.error_with_hints(
+                &self.loc(id),
+                format!(
+                    "cannot extract resource `{}` which is still borrowed",
+                    self.target.global_env().display(resource)
+                ),
+                "extracted here",
+                self.borrow_info(state, label, false, alive).into_iter(),
+            )
+        }
+    }
+
+    /// Process a return instruction.
+    fn return_(
+        &self,
+        state: &LifetimeState,
+        id: AttrId,
+        srcs: &[TempIndex],
+        alive: &LiveVarInfoAtCodeOffset,
+    ) {
+        for src in srcs {
+            if self.ty(*src).is_reference() {
+                // Need to check whether this reference is derived from a local which is not a
+                // a parameter
+                if let Some(label) = state.label_for_local(*src) {
+                    for root in state.roots(label) {
+                        match &state.node(root).location {
+                            MemoryLocation::Global(resource) => self.error_with_hints(
+                                &self.loc(id),
+                                format!(
+                                    "cannot return a reference derived from global `{}`",
+                                    self.target.global_env().display(resource)
+                                ),
+                                "returned here",
+                                self.borrow_info(state, root, false, alive).into_iter(),
+                            ),
+                            MemoryLocation::Local(local) => {
+                                if *local >= self.target.get_parameter_count() {
```

Comment:
Might be good to pull this logic out into `is_parameter` or some such to avoid exposing locals numbering scheme logic here.

Reply:
This is part of a very explicit contract in the model and used throughout the code. The first N locals/temporaries are parameters.

Reply:
"explicit" would imply documentation.  Abstraction would be nice, and would be self-documenting.